FWCore/Catalog/src/FileLocator.cc

```diff
@@ -187,6 +210,86 @@ namespace edm {
     }
   }
 
+  void FileLocator::init_da(std::string const& catAttr, unsigned iCatalog) {
+    Service<SiteLocalConfig> localconfservice;
+    std::string m_catAttr = catAttr;
+    if (m_catAttr.empty()) {
+      if (!localconfservice.isAvailable())
+        throw cms::Exception("FileCatalog", "edm::SiteLocalConfigService is not available");
+      if (iCatalog >= localconfservice->dataCatalogs().size())
+        throw cms::Exception("FileCatalog", "edm::FileLocator: Request nonexistence data catalog");
+      m_catAttr = localconfservice->dataCatalogs(1)[iCatalog];
+    }
+    std::vector<std::string> m_catAttrs;
+    boost::algorithm::split(m_catAttrs, m_catAttr, boost::is_any_of(std::string(",")));
+
+    std::string siteconfig_path = std::string(std::getenv("SITECONFIG_PATH"));
+    std::string m_filename_storage;
+    //not a cross site use local path given in SITECONFIG_PATH
+    if (m_catAttrs[0] == m_catAttrs[2]) {
+      //it is a site, use local path given in SITECONFIG_PATH
+      if (m_catAttrs[1] == m_catAttrs[0])
+        m_filename_storage = siteconfig_path + "/storage.json";
+      //it is a subsite, move one level up
+      else
+        m_filename_storage = siteconfig_path + "/../storage.json";
+    } else {  //cross site
+      //it is a site, move one level up
+      if (m_catAttrs[1] == m_catAttrs[0])
+        m_filename_storage = siteconfig_path + "/../" + m_catAttrs[2] + "/storage.json";
+      //it is a subsite, move two level up
+      else
+        m_filename_storage = siteconfig_path + "/../../" + m_catAttrs[2] + "/storage.json";
+    }
+
+    //now read json
+    pt::ptree json;
+    try {
+      boost::property_tree::read_json(m_filename_storage, json);
+    } catch (std::exception& e) {
+      throw cms::Exception("FileCatalog",
+                           "edm::FileCatalog:init_da Can not open storage.json (" + m_filename_storage +
+                               "). Check SITECONFIG_PATH and site-local-config.xml <data-access>");
+    }
+    //loop over sites
+    bool foundMatch(false);
+    BOOST_FOREACH (pt::ptree::value_type const& site, json) {
+      if (foundMatch)
+        break;
```

Comment:
Done. If site, volume and protocol are not found, it is OK to continue which results in an "empty" file locator which will be cached later. The point is that there many entries of combination of "site+volume+protocol" defined in site-local-config.xml. If one fails, the next one will be tried until success or exception if all are tried.

Reply:
I decided to enforce the matching of site, volume and protocol between <data-access> and storage.json. If site, volume and protocol defined in <data-access> can not be found in storage.json (due to typos or being neglected, which needs to be cleaned), the code will throw exception. This requires the site administrators to keep up-to-date and correct configurations. The exception helps to prevent fatal error below [1], which caused by mismatch definitions resulting in empty rules to interpret physical paths of input files.
======================
[1] ----- Begin Fatal Exception 12-Apr-2022 14:41:40 CDT-----------------------
An exception of category 'LogicalFileNameNotFound' occurred while
   [0] Constructing the EventProcessor
   [1] Constructing input source of type PoolSource
Exception Message:
RootFileSequenceBase::initTheFile()
Logical file name '/store/data/Run2016C/DoubleMuon/MINIAOD/17Jul2018-v1/00000/02B5E9A5-878D-E811-832D-0CC47A545096.root' was not found in the file catalog.
If you wanted a local file, you forgot the 'file:' prefix
before the file name in your configuration file.
----- End Fatal Exception -------------------------------------------------
Failure using test_sitelocalconfig_source_cfg_working.py: status 65